File: android/AgentShell/src/main/java/com/agentos/shell/tools/GoogleAuth.kt
package com.agentos.shell.tools

import android.content.Context
import android.net.Uri
import android.util.Base64
import android.util.Log
import androidx.browser.customtabs.CustomTabsIntent
import com.agentos.shell.BuildConfig
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Google sign-in for SlyOS, done the portable way: OAuth 2.0 Authorization Code + PKCE in a Chrome
 * Custom Tab. No Google Play Services, so it works on Samsung, de-Googled phones and the keyless
 * prebuilt APK. Each user signs into THEIR OWN Google account; tokens live only on this device.
 *
 * Scope is intentionally minimal — calendar.events (+ openid/email just to show which account is
 * connected). We never see the user's data: the phone talks to Google directly.
 */
object GoogleAuth {
    private const val TAG = "SlyOS"
    private const val PREF = "slyos_google"
    private const val AUTH_EP = "https://accounts.google.com/o/oauth2/v2/auth"
    private const val TOKEN_EP = "https://oauth2.googleapis.com/token"
    private const val SCOPE = "openid email https://www.googleapis.com/auth/calendar.events " +
        "https://www.googleapis.com/auth/gmail.readonly https://www.googleapis.com/auth/gmail.send"

    private fun prefs(ctx: Context) = ctx.getSharedPreferences(PREF, Context.MODE_PRIVATE)

    fun clientId(): String = BuildConfig.GOOGLE_OAUTH_CLIENT_ID
    fun configured(): Boolean = clientId().isNotBlank() && BuildConfig.GOOGLE_REDIRECT_SCHEME != "com.agentos.shell.noauth"
    private fun redirectUri(): String = BuildConfig.GOOGLE_REDIRECT_SCHEME + ":/oauth2redirect"

    fun isConnected(ctx: Context): Boolean = prefs(ctx).getString("refresh_token", "").orEmpty().isNotBlank()
    fun account(ctx: Context): String = prefs(ctx).getString("email", "").orEmpty()

    fun disconnect(ctx: Context) {
        prefs(ctx).edit().remove("refresh_token").remove("access_token")
            .remove("expiry").remove("email").remove("pkce_verifier").remove("state").apply()
    }

    // ---- PKCE helpers ----
    private fun randomUrlSafe(bytes: Int): String {
        val b = ByteArray(bytes); SecureRandom().nextBytes(b)
        return Base64.encodeToString(b, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    }
    private fun challenge(verifier: String): String {
        val d = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII))
        return Base64.encodeToString(d, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    }

    /** Begin sign-in: open Google's consent screen in a Custom Tab. Result returns via the redirect. */
    fun connect(ctx: Context): String {
        if (!configured()) return "Google isn't set up in this build yet."
        val verifier = randomUrlSafe(48)
        val state = randomUrlSafe(16)
        prefs(ctx).edit().putString("pkce_verifier", verifier).putString("state", state).apply()
        val url = Uri.parse(AUTH_EP).buildUpon()
            .appendQueryParameter("client_id", clientId())
            .appendQueryParameter("redirect_uri", redirectUri())
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("scope", SCOPE)
            .appendQueryParameter("code_challenge", challenge(verifier))
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("state", state)
            .appendQueryParameter("access_type", "offline")
            .appendQueryParameter("prompt", "consent")
            .build()
        return try {
            val tab = CustomTabsIntent.Builder().setShowTitle(true).build()
            tab.intent.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK)
            tab.launchUrl(ctx, url)
            ""
        } catch (e: Exception) {
            // Fallback: any browser.
            try {
                ctx.startActivity(android.content.Intent(android.content.Intent.ACTION_VIEW, url)
                    .addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK)); ""
            } catch (e2: Exception) { "Couldn't open a browser to sign in." }
        }
    }

    /** Handle the redirect (code or error). Exchanges the code for tokens. Returns ""=ok or an error. */
    fun handleRedirect(ctx: Context, data: Uri): String {
        val err = data.getQueryParameter("error")
        if (err != null) { Log.w(TAG, "google oauth error: $err"); return "Google sign-in was cancelled or denied." }
        val code = data.getQueryParameter("code") ?: return "No authorization code returned."
        val state = data.getQueryParameter("state")
        val savedState = prefs(ctx).getString("state", "")
        if (state != savedState) return "Sign-in state mismatch — please try again."
        val verifier = prefs(ctx).getString("pkce_verifier", "").orEmpty()
        if (verifier.isBlank()) return "Sign-in session expired — please try again."
        return exchangeCode(ctx, code, verifier)
    }

    private fun exchangeCode(ctx: Context, code: String, verifier: String): String {
        val form = "code=" + Uri.encode(code) +
            "&client_id=" + Uri.encode(clientId()) +
            "&redirect_uri=" + Uri.encode(redirectUri()) +
            "&grant_type=authorization_code" +
            "&code_verifier=" + Uri.encode(verifier)
        val (code2, body) = post(TOKEN_EP, form)
        if (code2 != 200) { Log.e(TAG, "token exchange $code2: $body"); return "Couldn't finish sign-in ($code2)." }
        return try {
            val o = JSONObject(body)
            val access = o.getString("access_token")
            val refresh = o.optString("refresh_token")
            val expiresIn = o.optLong("expires_in", 3600)
            val email = emailFromIdToken(o.optString("id_token"))
            val ed = prefs(ctx).edit()
            ed.putString("access_token", access)
            if (refresh.isNotBlank()) ed.putString("refresh_token", refresh)
            ed.putLong("expiry", System.currentTimeMillis() + (expiresIn - 60) * 1000)
            if (email.isNotBlank()) ed.putString("email", email)
            ed.remove("pkce_verifier").remove("state")
            ed.apply()
            ""
        } catch (e: Exception) { Log.e(TAG, "token parse failed", e); "Sign-in response was malformed." }
    }

    /** A valid access token, refreshing transparently if expired. Empty string if not connected. */
    fun accessToken(ctx: Context): String {
        val p = prefs(ctx)
        val cur = p.getString("access_token", "").orEmpty()
        val expiry = p.getLong("expiry", 0)
        if (cur.isNotBlank() && System.currentTimeMillis() < expiry) return cur
        val refresh = p.getString("refresh_token", "").orEmpty()
        if (refresh.isBlank()) return ""
        val form = "client_id=" + Uri.encode(clientId()) +
            "&grant_type=refresh_token&refresh_token=" + Uri.encode(refresh)
        val (code, body) = post(TOKEN_EP, form)
        if (code != 200) { Log.e(TAG, "refresh $code: $body"); return "" }
        return try {
            val o = JSONObject(body)
            val access = o.getString("access_token")
            val expiresIn = o.optLong("expires_in", 3600)
            p.edit().putString("access_token", access)
                .putLong("expiry", System.currentTimeMillis() + (expiresIn - 60) * 1000).apply()
            access
        } catch (e: Exception) { "" }
    }

    private fun emailFromIdToken(idToken: String): String {
        if (idToken.isBlank()) return ""
        return try {
            val parts = idToken.split(".")
            if (parts.size < 2) return ""
            val payload = String(Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP))
            JSONObject(payload).optString("email")
        } catch (e: Exception) { "" }
    }

    private fun post(endpoint: String, form: String): Pair<Int, String> {
        return try {
            val c = (URL(endpoint).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                connectTimeout = 20000; readTimeout = 20000
                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            }
            c.outputStream.use { it.write(form.toByteArray(Charsets.UTF_8)) }
            val code = c.responseCode
            val stream = if (code in 200..299) c.inputStream else c.errorStream
            val body = stream?.bufferedReader()?.use { it.readText() } ?: ""
            code to body
        } catch (e: Exception) { Log.e(TAG, "post failed", e); 0 to (e.message ?: "network error") }
    }
}

File: ios/SlyOS/Sources/SlyOSCompanion/GoogleAuth.swift
import Foundation
import AuthenticationServices
import CryptoKit

/// Google sign-in, ported from the Android `GoogleAuth`.
///
/// Same flow: authorization code with PKCE (S256) and **no client secret**, which is what makes it
/// safe to ship in an app binary at all. Each user signs into their own account and the tokens
/// never leave the device.
///
/// Two things differ from Android and both are deliberate:
///
/// * `ASWebAuthenticationSession` replaces the Custom Tab. It is the only API that shares Safari's
///   cookie jar, so someone already signed into Google is one tap from done — and it is the one
///   Apple accepts in review for third-party sign-in.
/// * Tokens live in the **Keychain**, not `UserDefaults`. A refresh token is a long-lived key to
///   the user's mail and calendar; `UserDefaults` is a plist that comes out in an unencrypted
///   backup. Android's `SharedPreferences` had the same weakness and this is the chance to fix it.
@Observable
final class GoogleAuth: NSObject {

    static let shared = GoogleAuth()

    private static let authEndpoint = "https://accounts.google.com/o/oauth2/v2/auth"
    private static let tokenEndpoint = "https://oauth2.googleapis.com/token"

    /// Exactly the scopes the Android build asks for, so one brain behaves the same on both phones.
    private static let scopes = [
        "openid", "email",
        "https://www.googleapis.com/auth/calendar.events",
        "https://www.googleapis.com/auth/gmail.readonly",
        "https://www.googleapis.com/auth/gmail.send",
        "https://www.googleapis.com/auth/documents",
        "https://www.googleapis.com/auth/presentations",
        "https://www.googleapis.com/auth/spreadsheets",
        "https://www.googleapis.com/auth/drive.file"
    ].joined(separator: " ")

    /// Whether the account is connected. Observable so the UI can react the moment it changes.
    private(set) var connectedEmail: String?
    var isConnected: Bool { connectedEmail != nil }

    /// The iOS OAuth client id, read from Info.plist (`GoogleOAuthClientID`).
    ///
    /// This must be an **iOS** client in Google Cloud Console — the Android client id will not work,
    /// because Google ties the redirect scheme to the client type.
    var clientID: String {
        Bundle.main.object(forInfoDictionaryKey: "GoogleOAuthClientID") as? String ?? ""
    }

    /// Google's iOS convention: the reversed client id is the URL scheme.
    private var redirectURI: String {
        let reversed = clientID.split(separator: ".").reversed().joined(separator: ".")
        return "\(reversed):/oauth2redirect"
    }

    var isConfigured: Bool { !clientID.isEmpty }

    private var session: ASWebAuthenticationSession?
    private let keychain = Keychain(service: "com.belto.slyos.google")

    private override init() {
        super.init()
        connectedEmail = keychain.string(for: "email")
    }

    // MARK: - Sign in

    enum AuthError: LocalizedError {
        case notConfigured, cancelled, denied(String), badResponse(Int, String), malformed

        var errorDescription: String? {
            switch self {
            case .notConfigured:
                "Google sign-in isn't set up in this build (no OAuth client id)."
            case .cancelled:
                "Sign-in was cancelled."
            case .denied(let why):
                "Google refused the sign-in: \(why)"
            case .badResponse(let code, _):
                "Couldn't finish sign-in (\(code))."
            case .malformed:
                "Google's sign-in response was malformed."
            }
        }
    }

    @MainActor
    func signIn() async throws {
        guard isConfigured else { throw AuthError.notConfigured }

        // PKCE: a fresh random verifier per attempt, and only its SHA-256 goes over the wire.
        let verifier = Self.randomVerifier()
        let challenge = Self.challenge(for: verifier)

        var components = URLComponents(string: Self.authEndpoint)!
        components.queryItems = [
            .init(name: "client_id", value: clientID),
            .init(name: "redirect_uri", value: redirectURI),
            .init(name: "response_type", value: "code"),
            .init(name: "scope", value: Self.scopes),
            .init(name: "code_challenge", value: challenge),
            .init(name: "code_challenge_method", value: "S256"),
            // Without both of these Google returns no refresh token on repeat sign-ins, and the
            // connection silently dies an hour later when the access token expires.
            .init(name: "access_type", value: "offline"),
            .init(name: "prompt", value: "consent")
        ]

        let callbackScheme = String(redirectURI.split(separator: ":").first ?? "")
        let code = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<String, Error>) in
            let s = ASWebAuthenticationSession(
                url: components.url!, callbackURLScheme: callbackScheme
            ) { url, error in
                if let error {
                    let cancelled = (error as? ASWebAuthenticationSessionError)?.code == .canceledLogin
                    cont.resume(throwing: cancelled ? AuthError.cancelled : error)
                    return
                }
                guard let url,
                      let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems
                else { return cont.resume(throwing: AuthError.malformed) }

                if let deny = items.first(where: { $0.name == "error" })?.value {
                    return cont.resume(throwing: AuthError.denied(deny))
                }
                guard let code = items.first(where: { $0.name == "code" })?.value else {
                    return cont.resume(throwing: AuthError.malformed)
                }
                cont.resume(returning: code)
            }
            s.presentationContextProvider = self
            // Use the existing Safari session so an already-signed-in user just taps Continue.
            s.prefersEphemeralWebBrowserSession = false
            self.session = s
            s.start()
        }

        try await exchange(code: code, verifier: verifier)
    }

    private func exchange(code: String, verifier: String) async throws {
        let body = [
            "code": code,
            "client_id": clientID,
            "redirect_uri": redirectURI,
            "grant_type": "authorization_code",
            "code_verifier": verifier
        ]
        let json = try await post(Self.tokenEndpoint, form: body)

        guard let access = json["access_token"] as? String else { throw AuthError.malformed }
        keychain.set(access, for: "access_token")
        if let refresh = json["refresh_token"] as? String, !refresh.isEmpty {
            keychain.set(refresh, for: "refresh_token")
        }
        let expiresIn = (json["expires_in"] as? Double) ?? 3600
        keychain.set(String(Date().timeIntervalSince1970 + expiresIn - 60), for: "expires_at")

        let email = (json["id_token"] as? String).flatMap(Self.emailFromIDToken) ?? "connected"
        keychain.set(email, for: "email")
        await MainActor.run { connectedEmail = email }
    }

    func signOut() {
        for k in ["access_token", "refresh_token", "expires_at", "email"] { keychain.remove(k) }
        connectedEmail = nil
    }

    // MARK: - Tokens

    /// A usable access token, refreshed transparently when it has expired.
    ///
    /// Every Google call goes through here rather than reading the stored token directly, because a
    /// token that expired sixty seconds ago fails with a 401 that looks exactly like "not connected".
    func accessToken() async throws -> String {
        guard let refresh = keychain.string(for: "refresh_token") else { throw AuthError.notConfigured }

        let expiry = keychain.string(for: "expires_at").flatMap(Double.init) ?? 0
        if let current = keychain.string(for: "access_token"), Date().timeIntervalSince1970 < expiry {
            return current
        }

        let json = try await post(Self.tokenEndpoint, form: [
            "client_id": clientID,
            "refresh_token": refresh,
            "grant_type": "refresh_token"
        ])
        guard let access = json["access_token"] as? String else { throw AuthError.malformed }
        keychain.set(access, for: "access_token")
        let expiresIn = (json["expires_in"] as? Double) ?? 3600
        keychain.set(String(Date().timeIntervalSince1970 + expiresIn - 60), for: "expires_at")
        return access
    }

    // MARK: - Plumbing

    private func post(_ endpoint: String, form: [String: String]) async throws -> [String: Any] {
        var req = URLRequest(url: URL(string: endpoint)!)
        req.httpMethod = "POST"
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        req.httpBody = form
            .map { "\($0.key)=\(Self.escape($0.value))" }
            .joined(separator: "&")
            .data(using: .utf8)
        req.timeoutInterval = 25

        let (data, response) = try await URLSession.shared.data(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw AuthError.badResponse(status, String(data: data, encoding: .utf8) ?? "")
        }
        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AuthError.malformed
        }
        return obj
    }

    private static func escape(_ s: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return s.addingPercentEncoding(withAllowedCharacters: allowed) ?? s
    }

    private static func randomVerifier() -> String {
        var bytes = [UInt8](repeating: 0, count: 64)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return base64URL(Data(bytes))
    }

    private static func challenge(for verifier: String) -> String {
        base64URL(Data(SHA256.hash(data: Data(verifier.utf8))))
    }

    /// PKCE and JWTs both use base64**url** — the plain base64 alphabet is rejected by Google.
    private static func base64URL(_ d: Data) -> String {
        d.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// Pull the email out of the id_token payload. The signature is not checked because the token
    /// came straight from Google's token endpoint over TLS; this is a display name, not a decision.
    private static func emailFromIDToken(_ jwt: String) -> String? {
        let parts = jwt.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var b64 = String(parts[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while b64.count % 4 != 0 { b64 += "=" }
        guard let data = Data(base64Encoded: b64),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return obj["email"] as? String
    }
}

extension GoogleAuth: ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }
            .first ?? ASPresentationAnchor()
    }
}
